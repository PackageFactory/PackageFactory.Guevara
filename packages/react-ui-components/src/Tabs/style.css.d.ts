export interface IStyleCss {
  'tabs': string;
  'tabs__content': string;
  'tabNavigation': string;
  'tabNavigation__item': string;
  'tabNavigation__item--isActive': string;
  'tabNavigation__itemBtn': string;
  'tabNavigation__itemBtnIcon': string;
  'tabNavigation__itemBtnIcon--hasLabel': string;
  'panel': string;
}

export const locals: IStyleCss;
