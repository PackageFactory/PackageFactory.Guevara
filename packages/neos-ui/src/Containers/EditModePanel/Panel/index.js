import React, {PropTypes, PureComponent} from 'react';
import HorizontalScroll from 'react-scroll-horizontal';

import Button from '@neos-project/react-ui-components/src/Button/index';
import I18n from '@neos-project/neos-ui-i18n';

export default class DimensionSwitcher extends PureComponent {
    static propTypes = {
        title: PropTypes.string.isRequired,
        className: PropTypes.string.isRequired,
        modes: PropTypes.array.isRequired,
        style: PropTypes.object.isRequired,
        current: PropTypes.string,
        currentMode: PropTypes.object,
        onPreviewModeClick: PropTypes.func.isRequired
    };

    render() {
        const {title, className, modes, current, currentMode, style, onPreviewModeClick} = this.props;

        const scrollConfiguration = {
            style: style.editModePanel__scrollable
        };

        return (
            <div className={className}>
                <p>{title} {currentMode && <b className={style.editModePanel__current}><I18n id={currentMode.title}/></b>}</p>
                <HorizontalScroll {...scrollConfiguration}>
                    {modes.map(previewMode => (
                        <Button
                            key={previewMode.id}
                            onClick={onPreviewModeClick(previewMode.id)}
                            style={previewMode.id === current ? 'brand' : null}
                            className={style.editModePanel__button}
                            >
                            <I18n id={previewMode.title}/>
                        </Button>
                    ))}
                </HorizontalScroll>
            </div>
        );
    }

}
