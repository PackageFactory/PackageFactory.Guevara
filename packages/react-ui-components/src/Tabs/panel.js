import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';

class Panel extends PureComponent {
    static propTypes = {
        /**
         * The title to be rendered within the navigation item of this Panel.
         */
        title: PropTypes.string,

        /**
         * An optional icon identifier which will be rendered within the navigation item of this Panel.
         */
        icon: PropTypes.string,

        /**
         * The children to render within the div node.
         */
        children: PropTypes.any.isRequired,

        /**
         * An optional css theme to be injected.
         */
        theme: PropTypes.shape({/* eslint-disable quote-props */
            'panel': PropTypes.string
        }).isRequired/* eslint-enable quote-props */
    }

    static displayName = 'Panel';

    render() {
        const {theme, children} = this.props;
        return (
            <div className={theme.panel}>{children}</div>
        );
    }
}

export default Panel;
