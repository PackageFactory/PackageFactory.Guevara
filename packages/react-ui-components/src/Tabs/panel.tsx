import React, {PureComponent} from 'react';

interface IPanelTheme {
    readonly 'panel': string;
}

interface IPanelProps {
    /**
     * The title to be rendered within the navigation item of this Panel.
     */
    readonly title: string;

    /**
     * An optional icon identifier which will be rendered within the navigation item of this Panel.
     */
    readonly icon?: string;

    /**
     * The children to render within the div node.
     */
    readonly children: any;

    /**
     * An optional css theme to be injected.
     */
    readonly theme?: IPanelTheme;
}

class Panel extends PureComponent<IPanelProps> {
    public static readonly displayName = 'Panel';

    public render(): JSX.Element {
        const {theme, children} = this.props;
        return (
            <div className={theme!.panel}>{children}</div>
        );
    }
}

export default Panel;
