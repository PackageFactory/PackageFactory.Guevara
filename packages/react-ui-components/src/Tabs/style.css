.tabs {
    composes: reset from './../reset.css';
    height: 100%;
}

.tabs__content {
    composes: reset from './../reset.css';
    height: calc(100% - 41px);
    overflow-y: auto;
}

.tabNavigation {
    composes: reset from './../reset.css';
    display: block;
    margin: 0;
    padding: 0;
    list-style: none;
    background: var(--colors-ContrastDarkest);
    border-bottom: 1px solid var(--colors-ContrastDark);

    &:before,
    &:after {
        content: '';
        display: table;
    }
    &:after {
        clear: both;
    }
}
.tabNavigation__item {
    composes: reset from './../reset.css';
    float: left;
    border-right: 1px solid var(--colors-ContrastDark);
    color: var(--colors-ContrastBrightest);
}
.tabNavigation__item--isActive {
    background-color: var(--colors-ContrastDarker);
    cursor: default;
    border-bottom: 1px solid var(--colors-PrimaryBlue);
    margin-bottom: -1px;

    > .tabNavigation__itemBtn {
        color: var(--colors-PrimaryBlue);

        &:hover {
            color: var(--colors-ContrastBrightest);
        }
    }
}
.tabNavigation__itemBtn {
    composes: reset from './../reset.css';
    display: inline-block;
    height: var(--spacing-GoldenUnit);
    padding: 0 var(--spacing-Full);
    line-height: var(--spacing-GoldenUnit);
    cursor: pointer;
    border: 0;
    background: 0;

    &:hover {
        background-color: var(--colors-PrimaryBlue);
    }

    &:focus {
        outline: 0;
    }
}
.tabNavigation__itemBtnIcon {
    color: currentColor;
}
.tabNavigation__itemBtnIcon--hasLabel {
    margin-right: var(--spacing-Half);
}

.panel {
    composes: reset from './../reset.css';
    overflow-y: auto;
}
