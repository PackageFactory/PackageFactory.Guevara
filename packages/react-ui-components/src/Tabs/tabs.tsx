import React, {PureComponent} from 'react';
import mergeClassNames from 'classnames';

import {PickDefaultProps} from '../../types';
import Panel from './panel.index';
import Icon from '../Icon';

export interface TabsProps {
    /**
     * The index of the active tab, defaults to 0.
     */
    readonly activeTab?: number;

    /**
     * An optional className to render on the wrapping div.
     */
    readonly className?: string;

    /**
     * The children panels to render.
     */
    readonly children: ReadonlyArray<React.ReactElement<any>>;

    /**
     * An optional css theme to be injected.
     */
    readonly theme?: TabsTheme;
}

interface TabsTheme extends TabMenuItemTheme {
    readonly 'tabs': string;
    readonly 'tabs__content': string;
    readonly 'tabNavigation': string;
}

export const tabsDefaultProps: PickDefaultProps<TabsProps, 'activeTab'> = {
    activeTab: 0
};

interface TabsState {
    readonly activeTab: number;
}

export default class Tabs extends PureComponent<TabsProps> {
    public static Panel = Panel;
    public state: TabsState = {
        activeTab: 0
    };

    public static defaultProps = tabsDefaultProps;

    public componentWillReceiveProps(newProps: TabsProps): void {
        const newActiveTab = newProps.activeTab;
        const {activeTab} = this.state;

        if (newActiveTab && newActiveTab !== activeTab) {
            this.setState({
                activeTab: newActiveTab
            });
        }
    }

    public getActiveTab(): number {
        // If activeTab is out of bounds, choose the first tab
        const {activeTab} = this.state;
        const childrenCount = React.Children.count(this.props.children);
        return childrenCount < activeTab + 1 ? 0 : activeTab;
    }

    public renderMenuItems(): JSX.Element {
        const {
            theme,
            children
        } = this.props;
        const activeTab = this.getActiveTab();

        const menuItems = children.map((panel, index) => (
            <TabMenuItem
                key={index}
                index={index}
                // tslint:disable-next-line:jsx-no-string-ref
                ref={`tab-${index}`}
                onClick={this.handleTabNavItemClick}
                isActive={activeTab === index}
                theme={theme!}
                title={panel.props.title}
                icon={panel.props.icon}
                tooltip={panel.props.tooltip}
            />
        ));

        return (
            <ul className={theme!.tabNavigation}>
                {menuItems}
            </ul>
        );
    }

    public handleTabNavItemClick = (index: number) => {
        this.setState({activeTab: index});
    }

    public renderPanels(): JSX.Element {
        const {theme, children} = this.props;
        const activeTab = this.getActiveTab();

        return (
            <div className={theme!.tabs__content}>
                {React.Children.map(children, (panel, index) => {
                    const isActive = activeTab === index;
                    const style = {
                        display: isActive ? 'block' : 'none'
                    };

                    return (
                        <div
                            key={index}
                            style={style}
                            role="tabpanel"
                            aria-hidden={isActive ? 'false' : 'true'}
                        >
                            {isActive && panel}
                        </div>
                    );
                })}
            </div>
        );
    }

    public render(): JSX.Element {
        const {theme, className} = this.props;
        const finalClassName = mergeClassNames(theme!.tabs, className);

        return (
            <div className={finalClassName} role="tablist">
                {this.renderMenuItems()}
                {this.renderPanels()}
            </div>
        );
    }
}

export interface TabMenuItemProps {
    /**
     * The index which will be handed over to the onClick handler.
     */
    index: number;

    /**
     * The title to render for the given Panel.
     */
    title?: string;

    /**
     * The click handler which will be called with the passed index as it's only argument.
     */
    onClick: (index: number) => void;

    /**
     * A boolean which controls if the rendered anchor is displayed as active or not.
     */
    isActive?: boolean;

    /**
     * An optional icon identifier, if one is passed, an Icon will be rendered besides the title.
     */
    icon?: string;

    /**
     * An optional tooltip for the given Panel.
     */
    tooltip?: string;

    theme?: TabMenuItemTheme;
}

interface TabMenuItemTheme {
    readonly 'tabNavigation__item': string;
    readonly 'tabNavigation__item--isActive': string;
    readonly 'tabNavigation__itemBtn': string;
    readonly 'tabNavigation__itemBtnIcon': string;
    readonly 'tabNavigation__itemBtnIcon--hasLabel': string;
}

export const tabMenuItemDefaultProps: PickDefaultProps<TabMenuItemProps, 'isActive'> = {
    isActive: false
};

// tslint:disable-next-line:max-classes-per-file
export class TabMenuItem extends PureComponent<TabMenuItemProps> {
    public static readonly defaultProps = tabMenuItemDefaultProps;

    public render(): JSX.Element {
        const {
            theme,
            isActive,
            index,
            icon,
            title,
            tooltip,
        } = this.props;
        const finalClassName = mergeClassNames(
            theme!.tabNavigation__item,
            {
                [theme!['tabNavigation__item--isActive']]: isActive,
            },
        );
        const finalIconClassName = mergeClassNames(
            theme!.tabNavigation__itemBtnIcon,
            {
                [theme!['tabNavigation__itemBtnIcon--hasLabel']]: title && title.length,
            },
        );

        return (
            <li className={finalClassName} role="presentation">
                <button
                    className={theme!.tabNavigation__itemBtn}
                    onClick={this.handleClick}
                    role="tab"
                    aria-selected={isActive ? 'true' : 'false'}
                    aria-controls={`section${index}`}
                    title={tooltip}
                >
                    {icon ? <Icon icon={icon} className={finalIconClassName}/> : null}
                    {title}
                </button>
            </li>
        );
    }

    private readonly handleClick = () => {
        this.props.onClick(this.props.index);
    }
}
