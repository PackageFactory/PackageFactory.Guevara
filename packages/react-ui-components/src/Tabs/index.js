import {themr} from 'react-css-themr';
import identifiers from './../identifiers.js';
import style from './style.css';
import Tabs from './tabs';
import Panel from './panel';
import TooltipContainer from '../TooltipContainer';

const ThemedTabs = themr(identifiers.tabs, style)(Tabs);

//
// Dependency injection
//
import injectProps from './../_lib/injectProps.js';
import Icon from './../Icon/index';

const FinalTabsComponent = injectProps({
    IconComponent: Icon,
    TooltipContainerComponent: TooltipContainer
})(ThemedTabs);
FinalTabsComponent.Panel = themr(identifiers.tabsPanel, style)(Panel);

export default FinalTabsComponent;
