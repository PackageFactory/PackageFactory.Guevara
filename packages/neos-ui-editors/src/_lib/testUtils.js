import {PureComponent, Children} from 'react';
import PropTypes from 'prop-types';
import {SynchronousRegistry, SynchronousMetaRegistry} from '@neos-project/neos-ui-extensibility/src/registry';

const globalRegistry = new SynchronousMetaRegistry();

// i18n Registry
class FakeI18NRegistry extends SynchronousRegistry {
    translate(key) {
        return key;
    }
}
globalRegistry.set('i18n', new FakeI18NRegistry());

// Data Source DataLoader
export const MockDataSourceDataLoader = {
    _currentPromise: null,
    _currentPromiseResolveFn: null,

    // API in tests
    reset() {
        this._currentPromise = null;
    },

    resolveCurrentPromise(result) {
        this._currentPromiseResolveFn(result);
        return this._currentPromise;
    },

    // API of DataLoader
    resolveValue() {
        if (!this._currentPromise) {
            this._currentPromise = new Promise(resolve => {
                this._currentPromiseResolveFn = resolve;
            });
        }
        return this._currentPromise;
    }
};
globalRegistry.set('dataLoaders', new SynchronousRegistry());
globalRegistry.get('dataLoaders').set('DataSources', MockDataSourceDataLoader);

// wrapper
export class WrapWithMockGlobalRegistry extends PureComponent {
    static propTypes = {
        children: PropTypes.element.isRequired
    };

    static childContextTypes = {
        globalRegistry: PropTypes.object.isRequired,
        configuration: PropTypes.object.isRequired
    };

    getChildContext() {
        const configuration = {};
        return {globalRegistry, configuration};
    }

    render() {
        return Children.only(this.props.children);
    }
}
