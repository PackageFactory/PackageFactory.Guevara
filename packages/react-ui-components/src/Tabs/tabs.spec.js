import React from 'react';
import * as enzyme from 'enzyme';
import {createStubComponent} from './../_lib/testUtils.js';
import Tabs, {TabMenuItem} from './tabs.js';
import Panel from './panel.js';

const IconComponent = createStubComponent();

const defaultProps = {
    theme: {
        'tabs': 'baseTabsClassName',
        'tabs__content': 'basTabsContentsClassName',
        'tabNavigation': 'baseTabsNavigationClassName',
        'tabNavigation__item': 'baseTabsNavigationItemClassName',
        'tabNavigation__item--isActive': 'activeTabsNavigationItemClassName',
        'tabNavigation__itemBtn': 'baseTabsNavigationItemBtnClassName',
        'tabNavigation__itemBtnIcon': 'baseTabsNavigationItemBtnIconClassName',
        'tabNavigation__itemBtnIcon--hasLabel': 'baseTabsNavigationItemBtnIconWithLabelClassName'
    },
    children: 'Foo children',
    IconComponent
};
const defaultPanelProps = {
    theme: {
        panel: 'panelBaseClassName'
    }
};
const shallow = props => {
    return enzyme.shallow(
        <Tabs {...defaultProps} {...props}>
            <Panel title="foo 1" icon="icon 1" {...defaultPanelProps}>Foo 1</Panel>
            <Panel title="foo 2" icon="icon 2" {...defaultPanelProps}>Foo 2</Panel>
            <Panel title="foo 3" icon="icon 3" {...defaultPanelProps}>Foo 3</Panel>
        </Tabs>
    );
};

test('should initialize with a state of {activeTab: 0}.', () => {
    const wrapper = shallow();

    expect(wrapper.state('activeTab')).toBe(0);
});
test('should render the themes "tabs" className.', () => {
    const wrapper = shallow();

    expect(wrapper.hasClass(defaultProps.theme.tabs)).toBeTruthy();
});
test('should render a "TabMenuItem" for each passed Panel.', () => {
    const wrapper = shallow();
    const items = wrapper.find(TabMenuItem);

    expect(items.length).toBe(3);
});
test('should pass a "isActive" prop to each "TabMenuItem".', () => {
    const wrapper = shallow();
    const items = wrapper.find(TabMenuItem);

    expect(items.at(0).prop('isActive')).toBe(true);
    expect(items.at(1).prop('isActive')).toBe(false);
    expect(items.at(2).prop('isActive')).toBe(false);
});
test('should pass a "onClick" prop to each "TabMenuItem" which matches the instances "handleTabNavItemClick" method.', () => {
    const wrapper = shallow();
    const items = wrapper.find(TabMenuItem);

    expect(items.at(0).prop('onClick')).toBe(wrapper.instance().handleTabNavItemClick);
});
test('should pass the Panels props to the "TabMenuItem" as well.', () => {
    const wrapper = shallow();
    const items = wrapper.find(TabMenuItem);

    expect(items.at(0).prop('title')).toBe('foo 1');
    expect(items.at(0).prop('icon')).toBe('icon 1');
});
test('should pass the static props of the "Tabs" component to the "TabMenuItem" as well.', () => {
    const wrapper = shallow();
    const items = wrapper.find(TabMenuItem);

    expect(items.at(0).prop('IconComponent')).toBe(defaultProps.IconComponent);
    expect(items.at(0).prop('theme')).toBe(defaultProps.theme);
});
test('should not attach the "tabNavigation__itemBtnIcon--hasLabel" className to the "TabMenuItem" if the Panels do not contain a titlte.', () => {
    const wrapper = enzyme.shallow(
        <Tabs {...defaultProps}>
            <Panel icon="icon 1" {...defaultPanelProps}>Foo 1</Panel>
            <Panel icon="icon 2" title="foo title" {...defaultPanelProps}>Foo 2</Panel>
            <Panel icon="icon 3" {...defaultPanelProps}>Foo 3</Panel>
        </Tabs>
    );
    const items = wrapper.find(TabMenuItem);

    expect(
        items.at(0).html().includes(defaultPanelProps.theme['tabNavigation__itemBtnIcon--hasLabel'])
    ).toBe(false);
});
test('should render each passed Panel wrapped in a div.', () => {
    const wrapper = shallow().find(`.${defaultProps.theme.tabs__content}`);
    const panels = wrapper.find(Panel);

    expect(wrapper.children().length).toBe(3);
    expect(panels.length).toBe(3);
});
test('should render a Panel wrapped in a div which is is hidden depending on the "activeTab" state.', () => {
    const wrapper = shallow().find(`.${defaultProps.theme.tabs__content}`);
    const divs = wrapper.children();

    expect(divs.at(0).html().includes('aria-hidden="false"')).toBeTruthy();
    expect(divs.at(0).html().includes('style="display:block;"')).toBeTruthy();
    expect(divs.at(1).html().includes('aria-hidden="true"')).toBeTruthy();
    expect(divs.at(1).html().includes('style="display:none;"')).toBeTruthy();
    expect(divs.at(2).html().includes('aria-hidden="true"')).toBeTruthy();
    expect(divs.at(2).html().includes('style="display:none;"')).toBeTruthy();
});
