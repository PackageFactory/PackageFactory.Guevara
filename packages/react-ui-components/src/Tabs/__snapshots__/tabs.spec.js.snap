// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs/> should render correctly. 1`] = `
<div
  className="baseTabsClassName"
  role="tablist"
>
  <ul
    className="baseTabsNavigationClassName"
  >
    <TabMenuItem
      IconComponent={[Function]}
      icon="icon 1"
      index={0}
      isActive={true}
      key="0/.0"
      onClick={[Function]}
      theme={
        Object {
          "tabNavigation": "baseTabsNavigationClassName",
          "tabNavigation__item": "baseTabsNavigationItemClassName",
          "tabNavigation__item--isActive": "activeTabsNavigationItemClassName",
          "tabNavigation__itemBtn": "baseTabsNavigationItemBtnClassName",
          "tabNavigation__itemBtnIcon": "baseTabsNavigationItemBtnIconClassName",
          "tabNavigation__itemBtnIcon--hasLabel": "baseTabsNavigationItemBtnIconWithLabelClassName",
          "tabs": "baseTabsClassName",
          "tabs__content": "basTabsContentsClassName",
        }
      }
      title="foo 1"
    />
    <TabMenuItem
      IconComponent={[Function]}
      icon="icon 2"
      index={1}
      isActive={false}
      key="1/.1"
      onClick={[Function]}
      theme={
        Object {
          "tabNavigation": "baseTabsNavigationClassName",
          "tabNavigation__item": "baseTabsNavigationItemClassName",
          "tabNavigation__item--isActive": "activeTabsNavigationItemClassName",
          "tabNavigation__itemBtn": "baseTabsNavigationItemBtnClassName",
          "tabNavigation__itemBtnIcon": "baseTabsNavigationItemBtnIconClassName",
          "tabNavigation__itemBtnIcon--hasLabel": "baseTabsNavigationItemBtnIconWithLabelClassName",
          "tabs": "baseTabsClassName",
          "tabs__content": "basTabsContentsClassName",
        }
      }
      title="foo 2"
    />
    <TabMenuItem
      IconComponent={[Function]}
      icon="icon 3"
      index={2}
      isActive={false}
      key="2/.2"
      onClick={[Function]}
      theme={
        Object {
          "tabNavigation": "baseTabsNavigationClassName",
          "tabNavigation__item": "baseTabsNavigationItemClassName",
          "tabNavigation__item--isActive": "activeTabsNavigationItemClassName",
          "tabNavigation__itemBtn": "baseTabsNavigationItemBtnClassName",
          "tabNavigation__itemBtnIcon": "baseTabsNavigationItemBtnIconClassName",
          "tabNavigation__itemBtnIcon--hasLabel": "baseTabsNavigationItemBtnIconWithLabelClassName",
          "tabs": "baseTabsClassName",
          "tabs__content": "basTabsContentsClassName",
        }
      }
      title="foo 3"
    />
  </ul>
  <div
    className="basTabsContentsClassName"
  >
    <div
      aria-hidden="false"
      key="0/.0"
      role="tabpanel"
      style={
        Object {
          "display": "block",
        }
      }
    >
      <Panel
        icon="icon 1"
        theme={
          Object {
            "panel": "panelBaseClassName",
          }
        }
        title="foo 1"
      >
        Foo 1
      </Panel>
    </div>
    <div
      aria-hidden="true"
      key="1/.1"
      role="tabpanel"
      style={
        Object {
          "display": "none",
        }
      }
    />
    <div
      aria-hidden="true"
      key="2/.2"
      role="tabpanel"
      style={
        Object {
          "display": "none",
        }
      }
    />
  </div>
</div>
`;
